libraries/node-core-library/src/Executable.ts

```diff
@@ -0,0 +1,419 @@
+          return resolvedNameWithExtension;
+        }
+      }
+    }
+
+    return undefined;
+  }
+
+  /**
+   * This is used when searching the shell PATH for an executable, to determine
+   * whether a match should be skipped or not.  If it returns true, this does not
+   * guarantee that the file can be successfully executed.
+   */
+  private static _canExecute(filePath: string, context: IExecutableContext): boolean {
+    if (!FileSystem.exists(filePath)) {
+      return false;
+    }
+
+    const fileExtension: string = path.extname(filePath);
+
+    if (os.platform() === 'win32') {
+      // Does the file have an executable file extension?
+      let matchFound: boolean = false;
+      for (const executableExtension of context.windowsExecutableExtensions) {
+        if (fileExtension.localeCompare(executableExtension) === 0) {
+          matchFound = true;
+          break;
+        }
+      }
+
+      if (fileExtension.localeCompare('.js') === 0) {
+        // The .js extension is special and always allowed since this is NodeJS.
+        // However, that file extension must be specified explitily unless it was added to PATHEXT
+        // (which is not usually the case).
+        matchFound = true;
+      }
+
+      if (!matchFound) {
+        return false;
+      }
+    } else {
+      // For Unix, check whether any of the POSIX execute bits are set, but don't
+      // require this for JavaSCript scripts (since this is NodeJS)
+      if (fileExtension.localeCompare('.js') !== 0) {
+        try {
+          // tslint:disable-next-line:no-bitwise
+          if ((FileSystem.getPosixModeBits(filePath) & PosixModeBits.AllExecute) === 0) {
+            return false; // not executable
+          }
+        } catch (error) {
+          // If we have trouble accessing the file, ignore the error and consider it "not executable"
+          // since that's what a shell would do
+        }
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Returns the list of folders where we will search for an executable,
+   * based on the PATH environment variable.
+   */
+  private static _getSearchFolders(context: IExecutableContext): string[] {
+
+    const pathList: string = context.environment.PATH || '';
+
+    const folders: string[] = [];
+
+    // Avoid processing duplicates
+    const seenPaths: Set<string> = new Set<string>();
+
+    if (os.platform() === 'win32') {
+      // On Window, the current directory is always tried first
+      folders.push(context.currentWorkingDirectory);
+      seenPaths.add(context.currentWorkingDirectory);
+    }
+
+    for (const splitPath of pathList.split(path.delimiter)) {
+      const trimmedPath: string = splitPath.trim();
+      if (trimmedPath !== '') {
+        if (!seenPaths.has(trimmedPath)) {
+          // Note that PATH is allowed to contain relative paths, which will be resolved
+          // relative to the current working directory.
```

Comment:
I'd learn towards breaking this feature on Windows, but leave a note here in case someone has an issue at some point.

Reply:
Alternatively we should warn if there's a relative path in the path variable.

---
In reply to: [208420814](https://github.com/Microsoft/web-build-tools/pull/764#discussion_r208420814) [](ancestors = 208420814)

Reply:
It actually seems like both Windows and BASH support this. I guess we just support it and expect nobody to ever use it.

---
In reply to: [208420897](https://github.com/Microsoft/web-build-tools/pull/764#discussion_r208420897) [](ancestors = 208420897,208420814)

Reply:
Let's do what the shells do, but issue a warning if we encounter a relative path in a PATH.

---
In reply to: [208421273](https://github.com/Microsoft/web-build-tools/pull/764#discussion_r208421273) [](ancestors = 208421273,208420897,208420814)